lib/liquid/context.rb

```diff
@@ -166,24 +146,9 @@ def find_variable(key, raise_on_not_found: true)
       index = @scopes.find_index { |s| s.key?(key) }
       scope = @scopes[index] if index
 
-      variable = nil
-
-      if scope.nil?
-        @environments.each do |e|
-          variable = lookup_and_evaluate(e, key, raise_on_not_found: raise_on_not_found)
```

Comment:
`lookup_and_evaluate` seems complete overkill here as it is always runs later once the scope has been decided anyway. So the approach is now to decide if the environment just contains the key and let the actually `lookup_and_evaluate` happen later (rather than running twice)

Reply:
FYI the entire `find_variable` method is overwritten by Liquid-C, so the changes that you are making here will only affect people running liquid without liquid-c. You might want to make the same changes in liquid-c if you think this is algorithmically faster.

Reply:
I'm having a hard time understanding why this change is equivalent (or if not equivalent then at least why it's ok that it's not).

The most obvious difference to me is that the previous version would evaluate Proc objects (and check their return value for being nil) when doing variable lookups, while your code doesn't seem to be doing that. So your code and the previous version of the code would potentially determine a different `scope` (yours will be happy with any Proc while the previous code was only happy with Proc objects that don't return nil).

Why does that not matter?

Reply:
What I don't get out of this is that the behaviour isn't really clear. For me the understanding of liquid is that when a variable is found you use it as it is defined. I don't actually get the usage of nil checks. Thinking about what your saying about the Proc calls I think something like the following makes more sense

```
      index = @scopes.find_index { |s| s.key?(key) || s[key].is_a?(Proc) }
      scope = @scopes[index] if index

      index = @environments.find_index { |s| s.key?(key) || s[key].is_a?(Proc) }
      scope = @environments[index] if index

      scope ||= {}
```
This means if a variable or proc is set use it as it was defined to handle the scope even if the response might be nil which is actually common.

That above snippet passes all tests as well.

Reply:
Isn't
```ruby
s.key?(key) || s[key].is_a?(Proc)
```
equivalent to just
```ruby
s.key?(key)
```
?

How can `s[key].is_a?(Proc)` be true without `s.key?(key)` already being true?

I'm not sure if evaluating procs makes sense when checking whether a variable is defined or not. I agree it seems unintuitive to me as well. But what matters here isn't so much which one is more reasonable but whether or not your change will break things in production that might rely on this behaviour.

Reply:
Yeah I agree everything still has to work and I'm sort of going backwards here figuring out whats going on and then I'll deep dive in to production usage to see if this break anything.

With out the Proc check this fails

```
  def test_hash_with_default_proc
    template = Template.parse(%(Hello {{ test }}))
    assigns = Hash.new { |h, k| raise "Unknown variable '#{k}'" }
    assigns['test'] = 'Tobi'
    assert_equal 'Hello Tobi', template.render!(assigns)
    assigns.delete('test')
    e = assert_raises(RuntimeError) do
      template.render!(assigns)
    end
    assert_equal "Unknown variable 'test'", e.message
  end
```

But I've now realised its not the proc check its the fact its accessing the variable which is what is triggering the Proc

Reply:
So I've ended up with the following based on what that test is trying to achieve.

```
      scope = (index = @scopes.find_index { |s| s.key?(key) || !s.default_proc.nil? }) && @scopes[index]
      scope ||= (index = @environments.find_index { |s| s.key?(key) || !s.default_proc.nil? }) && @environments[index]
      scope ||= @environments.last || @scopes.last
```

Reply:
`|| !s.default_proc.nil?` Ideally I want to confirm if this is actually useful or if it can be simplified

Reply:
So after going through and finding the actual history of this test. I came across exactly what I'm trying to resolve here. Liquid use to be exactly what I'm trying to do and I think just over time its become complicated for no reason. https://github.com/Shopify/liquid/commit/cfe3e6f3be197da14a1976fccb83cc1a33bb3add#diff-29a8bc9586d1e8466d9f3c03598d9f24

So this leaves me with
```
      scope = (index = @scopes.find_index { |s| s.key?(key)  }) && @scopes[index]
      scope ||= (index = @environments.find_index { |s| s.key?(key) }) && @environments[index]
      scope ||= {}

      variable = lookup_and_evaluate(scope, key, raise_on_not_found: raise_on_not_found).to_liquid
      variable.context = self if variable.respond_to?(:context=)

      variable
```

Reply:
Looks like this behaviour was introduced in https://github.com/Shopify/liquid/commit/167825aa925f0691b21d323f109b00f6332f72ad#diff-d97dbf3bc11085a7aa13ee71b9078b85R167

> further differentiate between environments and **only evaluate Procs once**

Reply:
So looking through this they aren't doing this because of `only evaluate Procs once` because this line was always going to achieve that. Its the addition of multiple environments is the reason they have done this https://github.com/Shopify/liquid/commit/167825aa925f0691b21d323f109b00f6332f72ad#diff-d97dbf3bc11085a7aa13ee71b9078b85L168

What I think has happened here is that they have included it because the original test would have failed when they were making this change. Because it fails for me now as well. But the original code was never intended to block it was just being used as a fallback and from the test it shows they were going to use as a way to detect missing variables https://github.com/Shopify/liquid/commit/cfe3e6f3be197da14a1976fccb83cc1a33bb3add#diff-d97dbf3bc11085a7aa13ee71b9078b85R162

I've also gone through usage in Shopify core and there is nothing that would be doing this, I've scanned all hash procs manually and there isn't anything that would justify this.

Also I'm just guessing https://github.com/Shopify/liquid/commits?author=candlerb wasn't actually Shopify but maybe external?

Reply:
The original design was so that the last scope item could be a hash with a fallback proc to handle missing variables. What it has turned in to though is something that doesn't really achieve anything but makes the original test pass.

Reply:
The actual proc support is still in lookup_and_evaluate and is still executed as always

Reply:
The only concern I still have is that the original code would mean that if one environment had a nil value it would still move on to the next environment which this doesn't maintain. I still need to look into where multiple environments are used as what could possibly have a nil value

Reply:
So going through how environments are actually used and for me this removes this providing any real risk.
```
drop.context = Context.new([drop, assigns], instance_assigns, registers, @rethrow_errors, @resource_limits)
```
```
 Context.new([args.shift, assigns], instance_assigns, registers, @rethrow_errors, @resource_limits)
```
Going through Shopify I can't find a single use of assigns anyway so this environment must always be empty, unless I'm missing something. If that environment is always empty then there isn't any risk here.